gson/src/main/java/com/google/gson/internal/reflect/ReflectionHelper.java

```diff
@@ -65,10 +82,131 @@ public static String tryMakeAccessible(Constructor<?> constructor) {
     }
   }
 
-  public static RuntimeException createExceptionForUnexpectedIllegalAccess(IllegalAccessException exception) {
+  public static boolean isRecord(Class<?> raw) {
+    return RECORD_HELPER.isRecord(raw);
+  }
+
+  public static String[] recordFields(Class<?> raw) {
+    return RECORD_HELPER.recordFields(raw);
+  }
+
+  public static RuntimeException createExceptionForUnexpectedIllegalAccess(
+      IllegalAccessException exception) {
     throw new RuntimeException("Unexpected IllegalAccessException occurred (Gson " + GsonBuildConfig.VERSION + "). "
         + "Certain ReflectionAccessFilter features require Java >= 9 to work correctly. If you are not using "
         + "ReflectionAccessFilter, report this to the Gson maintainers.",
         exception);
   }
+
+  public static <T> Constructor<T> getCanonicalRecordConstructor(Class<T> raw) {
+    return RECORD_HELPER.getCanonicalRecordConstructor(raw);
+  }
+
+  /**
+   * Internal abstraction over reflection when Records are supported.
+   */
+  private abstract static class RecordHelper {
+    abstract boolean isRecord(Class<?> clazz);
+
+    abstract String[] recordFields(Class<?> clazz);
+
+    abstract <T> Constructor<T> getCanonicalRecordConstructor(Class<T> raw);
+  }
+
+  private static class RecordSupportedHelper extends RecordHelper {
+    private final Method isRecord;
+    private final Method getRecordComponents;
+    private final Method getName;
+    private final Method getType;
+
+    private RecordSupportedHelper() throws NoSuchMethodException {
+      isRecord = Class.class.getMethod("isRecord");
+      getRecordComponents = Class.class.getDeclaredMethod("getRecordComponents");
+      Class<?> recordComponentType = getRecordComponents.getReturnType().getComponentType();
+      getName = recordComponentType.getDeclaredMethod("getName");
+      getType = recordComponentType.getDeclaredMethod("getType");
+    }
+
+    @Override
+    boolean isRecord(Class<?> raw) {
+      try {
+        return boolean.class.cast(isRecord.invoke(raw));
+      } catch (IllegalAccessException e) {
+        throw createExceptionForUnexpectedIllegalAccess(e);
```

Comment:
`createExceptionForUnexpectedIllegalAccess` is only intended for exceptions which occur when trying to access members of user provided classes. Its exception message mentions `ReflectionAccessFilter` which does not apply here.

Would probably be best to merge this with the `catch (InvocationTargetException ...` below instead.

Reply:
This has been changed.